File: javascript/webhooks-nodejs/createWebhook.js
const axios = require("axios");
const fs = require("node:fs");
const path = require("node:path");
require("dotenv").config();

// File path for the .env file
const envFilePath = path.join(__dirname, ".env");

// Helper function to read X-Hook-Secret from the .env file
function getXHookSecret() {
  const envContent = fs.readFileSync(envFilePath, "utf8");
  const match = envContent.match(/X_HOOK_SECRET=(.*)/);
  return match ? match[1] : "";
}

// Async function to delete the webhook
// Documentation: https://developers.asana.com/reference/deletewebhook
async function deleteWebhook(webhookId, personalAccessToken) {
  const url = `https://app.asana.com/api/1.0/webhooks/${webhookId}`;

  try {
    await axios.delete(url, {
      headers: {
        Authorization: `Bearer ${personalAccessToken}`,
      },
    });
    console.log(`Webhook ${webhookId} deleted due to mismatch.`);
  } catch (error) {
    console.error(
      "Error deleting webhook:",
      error.response ? error.response.data : error.message
    );
  }
}

// Async function to create the webhook
// Documentation: https://developers.asana.com/reference/createwebhook
async function createWebhook(targetUri, objectId, filter, resourceType) {
  const personalAccessToken = process.env.PAT;
  const url = "https://app.asana.com/api/1.0/webhooks";

  try {
    const response = await axios.post(
      url,
      {
        data: {
          resource: objectId,
          target: targetUri,
          filters: [{ action: filter, resource_type: resourceType }],
        },
      },
      {
        headers: {
          Authorization: `Bearer ${personalAccessToken}`,
        },
      }
    );

    const xHookSecret = response.data["X-Hook-Secret"];
    const storedSecret = getXHookSecret();

    // Compare the secrets. This check is not solely for detecting mutations in the stored secret.
    // Rather, it ensures that the original secret was valid and corresponds to what Asana expects.
    // Read more about validation during the webhook handshake: https://developers.asana.com/docs/webhooks-guide#the-webhook-handshake
    if (xHookSecret !== storedSecret) {
      console.error(
        `X-Hook-Secrets do not match! Potential security issue. Deleting webhook with GID ${response.data.data.gid}.`
      );
      // Delete the webhook if the secrets don't match
      await deleteWebhook(response.data.data.gid, personalAccessToken);
      return;
    }

    console.log("Webhook created successfully!");
    console.log(
      `The GID of the newly-created webhook is: ${response.data.data.gid}.`
    );
    console.log(
      `The X-Hook-Secret from Asana's '201 Created' response is: ${xHookSecret}.`
    );
  } catch (error) {
    console.error(
      "Error creating webhook:",
      error.response ? error.response.data : error.message
    );
  }
}

// TODO: Replace these values with your target URI, object ID, filter, and resource type
const targetUri = "<YOUR_URL_HERE>/receiveWebhook"; // The webhook server's public endpoint for receiving webhooks (view README for an example)
const objectId = "<OBJECT_ID_HERE>"; // The Asana object ID you want to track (e.g., a task gid)
const filter = "changed"; // The action to filter for (documentation: https://developers.asana.com/docs/webhooks-guide#actions)
const resourceType = "task"; // Specify the resource type (e.g., task, project)

createWebhook(targetUri, objectId, filter, resourceType);

File: python/webhooks_v5/create_webhook.py
import os
import asana
from asana.rest import ApiException
from dotenv import load_dotenv
from pprint import pprint

# Load environment variables (for stored PAT and X-Hook-Secret)
load_dotenv()

# Configure Asana client
configuration = asana.Configuration()
configuration.access_token = os.getenv('PAT')
api_client = asana.ApiClient(configuration)

# Create an instance of the Webhooks API
webhooks_api_instance = asana.WebhooksApi(api_client)

# Helper function to read X-Hook-Secret from the .env file
def get_x_hook_secret():
    with open('.env', 'r') as file:
        for line in file:
            if line.startswith('X_HOOK_SECRET='):
                return line.split('=')[1].strip()
    return ""

# Function to delete the webhook
# Documentation: https://github.com/Asana/python-asana/blob/v5.0.11/docs/WebhooksApi.md#delete_webhook
def delete_webhook(webhook_gid):
    try:
        api_response = webhooks_api_instance.delete_webhook(webhook_gid)
        print(f'Webhook {webhook_gid} deleted due to mismatch.')
    except ApiException as e:
        print(f"Exception when calling WebhooksApi->delete_webhook: {e}\n")

# Function to create the webhook
# Documentation: https://github.com/Asana/python-asana/blob/v5.0.11/docs/WebhooksApi.md#create_webhook
def create_webhook(target_uri, object_id, filter, resource_type):
    body = {
        "data": {
            "resource": object_id,
            "target": target_uri,
            "filters": [{"action": filter, "resource_type": resource_type}],
        }
    }

    # Documentation: https://developers.asana.com/docs/inputoutput-options
    opts = {}

    try:
        # Establish a webhook
        # Note: This request disables default pagination behavior
        # (see https://github.com/Asana/python-asana?tab=readme-ov-file#disabling-default-pagination-behaviour)
        api_response = webhooks_api_instance.create_webhook(body, opts, full_payload=True)

        print("The complete API response for POST /webhooks is below:")
        pprint(api_response)

        webhook_gid = api_response["data"]["gid"]
        
        x_hook_secret = api_response.get('X-Hook-Secret')
        stored_secret = get_x_hook_secret()

        if x_hook_secret != stored_secret:
            print(f'X-Hook-Secrets do not match! Deleting webhook with GID {webhook_gid}')
            delete_webhook(api_response["data"]["gid"])
        else:
            print("Webhook created successfully!")
            print(f'The GID of the newly-created webhook is: {webhook_gid}')
            print(f'The X-Hook-Secret from Asana\'s response is: {x_hook_secret}')

    except ApiException as e:
        print(f"Exception when calling WebhooksApi->create_webhook: {e}\n")

# TODO: Replace these values with your target URI, object ID, filter, and resource type
target_uri = '<YOUR_URL_HERE>/receive_webhook'  # The webhook server's public endpoint
object_id = '<OBJECT_ID_HERE>'  # The Asana object ID you want to track (e.g., a task gid)
filter = 'changed'  # The action to filter for
resource_type = 'task'  # Specify the resource type

create_webhook(target_uri, object_id, filter, resource_type)
